File: internal/agent/protocol_test.go
package agent

import (
	"bytes"
	"encoding/binary"
	"encoding/json"
	"errors"
	"io"
	"strings"
	"testing"
)

func roundTrip(t *testing.T, message Message) Message {
	t.Helper()
	var buffer bytes.Buffer
	if err := WriteMessage(&buffer, message); err != nil {
		t.Fatalf("WriteMessage: %v", err)
	}
	got, err := ReadMessage(&buffer)
	if err != nil {
		t.Fatalf("ReadMessage: %v", err)
	}
	return got
}

func TestRoundTrip_Hello(t *testing.T) {
	got := roundTrip(t, Hello("0.0.1", "android", "in.okcredit.merchant"))
	if got.Type != MessageTypeHello || got.Version != "0.0.1" || got.Platform != "android" || got.AppPackage != "in.okcredit.merchant" {
		t.Fatalf("hello round-trip failed: %+v", got)
	}
	if got.ProtocolVersion != ProtocolVersion {
		t.Errorf("protocol_version: got %d, want %d", got.ProtocolVersion, ProtocolVersion)
	}
}

func TestRoundTrip_PauseResume(t *testing.T) {
	for _, builder := range []func(uint64) Message{Pause, Resume} {
		got := roundTrip(t, builder(42))
		if got.ID != 42 {
			t.Errorf("id round-trip failed: %+v", got)
		}
	}
}

func TestRoundTrip_State(t *testing.T) {
	snapshots := map[string]json.RawMessage{
		"screen":         json.RawMessage(`"customer_ledger"`),
		"ledger.balance": json.RawMessage(`1500`),
		"is_signed_in":   json.RawMessage(`true`),
	}
	got := roundTrip(t, State(7, snapshots))
	if got.Type != MessageTypeState || got.ID != 7 {
		t.Fatalf("state envelope wrong: %+v", got)
	}
	if string(got.Snapshots["screen"]) != `"customer_ledger"` {
		t.Errorf("screen snapshot wrong: %s", got.Snapshots["screen"])
	}
	if string(got.Snapshots["ledger.balance"]) != `1500` {
		t.Errorf("balance snapshot wrong: %s", got.Snapshots["ledger.balance"])
	}
}

func TestRoundTrip_ExtractResult(t *testing.T) {
	got := roundTrip(t, ExtractResult(1, "ledger.balance", json.RawMessage(`2500`), ""))
	if got.Extractor != "ledger.balance" || string(got.Result) != `2500` {
		t.Fatalf("extract result round-trip failed: %+v", got)
	}

	failed := roundTrip(t, ExtractResult(2, "ledger.balance", nil, "no active customer"))
	if failed.Error != "no active customer" {
		t.Errorf("extract error round-trip failed: %+v", failed)
	}
}

func TestRoundTrip_Goodbye(t *testing.T) {
	got := roundTrip(t, Goodbye("app terminated"))
	if got.Type != MessageTypeGoodbye || got.Reason != "app terminated" {
		t.Fatalf("goodbye round-trip failed: %+v", got)
	}
}

func TestWriteMessage_FrameFormat(t *testing.T) {
	var buffer bytes.Buffer
	if err := WriteMessage(&buffer, Pause(99)); err != nil {
		t.Fatal(err)
	}
	raw := buffer.Bytes()
	if len(raw) < 4 {
		t.Fatalf("frame too short: %d bytes", len(raw))
	}
	length := binary.BigEndian.Uint32(raw[:4])
	if int(length) != len(raw)-4 {
		t.Errorf("header length %d mismatches payload length %d", length, len(raw)-4)
	}
	if !strings.Contains(string(raw[4:]), `"type":"PAUSE"`) {
		t.Errorf("payload does not contain PAUSE type: %s", raw[4:])
	}
}

func TestReadMessage_ShortReaderReturnsEOF(t *testing.T) {
	_, err := ReadMessage(bytes.NewReader(nil))
	if !errors.Is(err, io.EOF) {
		t.Errorf("expected EOF on empty reader, got %v", err)
	}
}

func TestReadMessage_OversizedFrameRejected(t *testing.T) {
	var header [4]byte
	binary.BigEndian.PutUint32(header[:], uint32(MaxFrameSize+1))
	_, err := ReadMessage(bytes.NewReader(header[:]))
	if err == nil || !strings.Contains(err.Error(), "exceeds maximum") {
		t.Errorf("expected oversized-frame error, got %v", err)
	}
}

func TestReadMessage_MissingTypeRejected(t *testing.T) {
	var buffer bytes.Buffer
	payload := []byte(`{"id":1}`)
	var header [4]byte
	binary.BigEndian.PutUint32(header[:], uint32(len(payload)))
	buffer.Write(header[:])
	buffer.Write(payload)

	_, err := ReadMessage(&buffer)
	if err == nil || !strings.Contains(err.Error(), "missing type") {
		t.Errorf("expected missing-type error, got %v", err)
	}
}

func TestWriteMessage_StreamsMultipleFrames(t *testing.T) {
	var buffer bytes.Buffer
	messages := []Message{
		Hello("v", "android", "com.x"),
		Pause(1),
		State(1, map[string]json.RawMessage{"x": json.RawMessage(`42`)}),
		Resume(1),
		Goodbye("done"),
	}
	for _, message := range messages {
		if err := WriteMessage(&buffer, message); err != nil {
			t.Fatal(err)
		}
	}
	for index, want := range messages {
		got, err := ReadMessage(&buffer)
		if err != nil {
			t.Fatalf("frame %d: %v", index, err)
		}
		if got.Type != want.Type {
			t.Errorf("frame %d: got type %q, want %q", index, got.Type, want.Type)
		}
	}
}

File: sdk/android/src/test/kotlin/dev/uatu/sdk/ProtocolTest.kt
package dev.uatu.sdk

import java.io.ByteArrayInputStream
import java.io.ByteArrayOutputStream
import java.io.DataInputStream
import java.io.EOFException
import java.io.IOException
import org.junit.Assert.assertArrayEquals
import org.junit.Assert.assertEquals
import org.junit.Assert.assertNotNull
import org.junit.Assert.assertNull
import org.junit.Assert.assertThrows
import org.junit.Assert.assertTrue
import org.junit.Test

class ProtocolTest {

    private fun roundTrip(message: Message): Message {
        val output = ByteArrayOutputStream()
        Protocol.write(output, message)
        return Protocol.read(ByteArrayInputStream(output.toByteArray()))
    }

    @Test fun roundTripHello() {
        val got = roundTrip(Message.hello("0.0.1", "android", "in.okcredit.merchant"))
        assertEquals(MessageType.HELLO, got.type)
        assertEquals(Protocol.PROTOCOL_VERSION, got.protocolVersion)
        assertEquals("0.0.1", got.version)
        assertEquals("android", got.platform)
        assertEquals("in.okcredit.merchant", got.appPackage)
    }

    @Test fun roundTripPauseResume() {
        val pause = roundTrip(Message.pause(42))
        assertEquals(MessageType.PAUSE, pause.type)
        assertEquals(42L, pause.id)

        val resume = roundTrip(Message.resume(43))
        assertEquals(MessageType.RESUME, resume.type)
        assertEquals(43L, resume.id)
    }

    @Test fun roundTripState() {
        val snapshots = mapOf<String, Any?>(
            "screen" to "customer_ledger",
            "ledger.balance" to 1500,
            "is_signed_in" to true,
        )
        val got = roundTrip(Message.state(7, snapshots))
        assertEquals(MessageType.STATE, got.type)
        assertEquals(7L, got.id)
        assertNotNull(got.snapshots)
        assertEquals("customer_ledger", got.snapshots!!["screen"])
        assertEquals(1500, got.snapshots["ledger.balance"])
        assertEquals(true, got.snapshots["is_signed_in"])
    }

    @Test fun roundTripExtractResult() {
        val ok = roundTrip(Message.extractResult(1, "ledger.balance", 2500))
        assertEquals("ledger.balance", ok.extractor)
        assertEquals(2500, ok.result)
        assertNull(ok.error)

        val failed = roundTrip(Message.extractResult(2, "ledger.balance", null, "no active customer"))
        assertEquals("no active customer", failed.error)
        assertNull(failed.result)
    }

    @Test fun roundTripGoodbye() {
        val got = roundTrip(Message.goodbye("app terminated"))
        assertEquals(MessageType.GOODBYE, got.type)
        assertEquals("app terminated", got.reason)
    }

    @Test fun frameFormatIsBigEndianLengthPlusJson() {
        val output = ByteArrayOutputStream()
        Protocol.write(output, Message.pause(99))
        val raw = output.toByteArray()
        assertTrue("frame must have 4-byte header plus body", raw.size > 4)
        val length = DataInputStream(ByteArrayInputStream(raw.copyOfRange(0, 4))).readInt()
        assertEquals(raw.size - 4, length)
        val payload = String(raw.copyOfRange(4, raw.size), Charsets.UTF_8)
        assertTrue("payload should contain PAUSE type, got $payload", payload.contains("\"type\":\"PAUSE\""))
    }

    @Test fun emptyReaderThrowsEof() {
        assertThrows(EOFException::class.java) {
            Protocol.read(ByteArrayInputStream(ByteArray(0)))
        }
    }

    @Test fun oversizedFrameRejected() {
        val header = ByteArray(4)
        val tooBig = Protocol.MAX_FRAME_SIZE + 1
        header[0] = (tooBig ushr 24).toByte()
        header[1] = (tooBig ushr 16).toByte()
        header[2] = (tooBig ushr 8).toByte()
        header[3] = tooBig.toByte()
        val error = assertThrows(IOException::class.java) {
            Protocol.read(ByteArrayInputStream(header))
        }
        assertTrue("expected size error, got: ${error.message}", error.message!!.contains("exceeds maximum"))
    }

    @Test fun missingTypeRejected() {
        val payload = "{\"id\":1}".toByteArray(Charsets.UTF_8)
        val header = ByteArray(4)
        header[0] = (payload.size ushr 24).toByte()
        header[1] = (payload.size ushr 16).toByte()
        header[2] = (payload.size ushr 8).toByte()
        header[3] = payload.size.toByte()
        val frame = header + payload
        val error = assertThrows(IOException::class.java) {
            Protocol.read(ByteArrayInputStream(frame))
        }
        assertTrue("expected missing-type error, got: ${error.message}", error.message!!.contains("missing type"))
    }

    @Test fun streamsMultipleFrames() {
        val messages = listOf(
            Message.hello("v", "android", "com.x"),
            Message.pause(1),
            Message.state(1, mapOf("x" to 42)),
            Message.resume(1),
            Message.goodbye("done"),
        )
        val output = ByteArrayOutputStream()
        for (message in messages) Protocol.write(output, message)

        val input = ByteArrayInputStream(output.toByteArray())
        for (want in messages) {
            val got = Protocol.read(input)
            assertEquals(want.type, got.type)
        }
    }

    @Test fun sharedWireFormatMatchesGoEncoder() {
        // Fixture encoded by the Go side (see internal/agent/protocol.go).
        // Ensures both encoders agree on field names and ordering conventions.
        val output = ByteArrayOutputStream()
        Protocol.write(output, Message.hello("0.0.1", "android", "com.x"))
        val payload = String(output.toByteArray().copyOfRange(4, output.size()), Charsets.UTF_8)
        assertTrue(payload.contains("\"type\":\"HELLO\""))
        assertTrue(payload.contains("\"protocol_version\":1"))
        assertTrue(payload.contains("\"version\":\"0.0.1\""))
        assertTrue(payload.contains("\"platform\":\"android\""))
        assertTrue(payload.contains("\"app_package\":\"com.x\""))
    }

    @Test fun bytesAreConsumedInOrder() {
        // Regression guard: a second read shouldn't see stale bytes.
        val output = ByteArrayOutputStream()
        Protocol.write(output, Message.pause(1))
        Protocol.write(output, Message.pause(2))
        val bytes = output.toByteArray()
        val input = ByteArrayInputStream(bytes)
        assertEquals(1L, Protocol.read(input).id)
        assertEquals(2L, Protocol.read(input).id)
        // ByteArrayInputStream should be drained.
        val leftover = ByteArray(bytes.size)
        val remaining = input.read(leftover)
        assertEquals(-1, remaining)
        assertArrayEquals(ByteArray(bytes.size), leftover)
    }
}
